### PR TITLE
[vllm, rollout, cfg, doc] feat: Accelerate RL rollouts with EAGLE/EAGLE3 speculative decoding

### PR DESCRIPTION
**Pull request by alekseymalakhov11**
### What does this PR do?

> Add inference-only speculative decoding for vLLM rollouts using EAGLE / EAGLE3 draft models. This PR adds the vLLM rollout/config wiring for the new speculative decoding path, reload handling for the draft model during weight sync, rollout metrics for speculative decoding, and documentation.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: [rollout] feat: support eagle3 speculative decode in rollout
#5509

- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Screenshot 2026-04-08 at 16 27 09
Screenshot 2026-04-08 at 16 26 05
Screenshot 2026-04-08 at 16 25 14

We used `qwen3-8b` together with `RedHatAI/Qwen3-8B-speculator.eagle3` on `verl-team/lighteval-MATH-preprocessed`.

In our experiments, the final reward is comparable to the baseline. The screenshots also highlights two implementation details that are important for making this work correctly in RL rollout: reloading the draft-model weights after actor weight sync, and rebuilding the RoPE `cos_sin cache` after reload. Also it shows speculative decoding metrics such as acceptance rate and mean acceptance length so the behavior and performance of EAGLE/EAGLE3 can be analyzed directly during training.

In some settings, performance of draft model may degrade (acceptance length drops). Even with that limitation, this change opens a useful direction for more practical research on speculative decoding in RL for LLMs, and makes it easier to explore follow-up ideas such as retraining or adapting the draft model.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by wuxibin89** [comment]
`verl-team/lighteval-MATH-preprocessed` is not enough to verify convergence, please verify [BytedTsinghua-SIA/DAPO-Math-17k](https://huggingface.co/datasets/BytedTsinghua-SIA/DAPO-Math-17k) and [nvidia/OpenMathReasoning](https://huggingface.co/datasets/nvidia/OpenMathReasoning)

**Comment by alekseymalakhov11** [comment]
> `verl-team/lighteval-MATH-preprocessed` is not enough to verify convergence, please verify [BytedTsinghua-SIA/DAPO-Math-17k](https://huggingface.co/datasets/BytedTsinghua-SIA/DAPO-Math-17k) and [nvidia/OpenMathReasoning](https://huggingface.co/datasets/nvidia/OpenMathReasoning)


Hi @wuxibin89 👋
Thanks for your feedback!

I ran a few additional experiments. From what I can see so far, the behavior looks stable, training curves for reward, entropy, response length, and gradient norm appear consistent. 
I can keep the runs going longer if needed, but at this point I do not expect the overall picture to change significantly.

I also tested another dataset with the same base model (`qwen3-8b`) and the `RedHatAI/Qwen3-8B-speculator.eagle3` draft model:

Dataset: `agentica-org/DeepScaleR-Preview-Dataset`

photo_1_2026-04-09_18-32-52
Screenshot 2026-04-09 at 18 16 03
Screenshot 2026-04-09 at 18 16 16
Screenshot 2026-04-09 at 18 16 28

Dataset: `BytedTsinghua-SIA/DAPO-Math-17k`

Screenshot 2026-04-09 at 18 18 29
Screenshot 2026-04-09 at 18 18 37
Screenshot 2026-04-09 at 18 18 43
Screenshot 2026-04-09 at 18 18 55

These runs used the default `ppo_config` with the Megatron engine and the following training parameters:

```yaml
data.train_batch_size: 256
data.max_prompt_length: 2048
data.max_response_length: 6144
actor_rollout_ref.actor.ppo_mini_batch_size: 16

rollout.tensor_model_parallel_size: 1
rollout.gpu_memory_utilization: 0.8
rollout.name: vllm
rollout.temperature: 1.0

rollout.speculative_decoding.method: eagle3
rollout.speculative_decoding.num_draft_tokens: 4
rollout.speculative_decoding.draft_tensor_parallel_size: 1
```

I am also planning to continue working on accelerating RL rollouts with EAGLE-based speculative decoding and hopefully improve the speedup further. 

Still, I think the current results already show a useful contribution: with basic open-source speculative models, the rollout phase can already be accelerated by roughly 15% with minimal additional setup.

**Comment by alekseymalakhov11** [comment]
> I can keep the runs going longer if needed, but at this point I do not expect the overall picture to change significantly.

Here is a longer version of the previous runs. As expected, the results remain consistent and everything looks stable.

Dataset: `agentica-org/DeepScaleR-Preview-Dataset`
Screenshot 2026-04-10 at 19 57 36
Screenshot 2026-04-10 at 19 57 40
Screenshot 2026-04-10 at 19 57 47



Dataset: `BytedTsinghua-SIA/DAPO-Math-17k`
Screenshot 2026-04-10 at 19 56 39
Screenshot 2026-04-10 at 19 56 46
Screenshot 2026-04-10 at 19 57 00

**Comment by wuxibin89** [comment]
@tpx818 is working on a co-training draft model version.

**Comment by tpx818** [comment]
> DAPO-Math-17k

When using the DAPO-Math-17k dataset, we observed a reward collapse around step 100-120. If you try training for more steps, this phenomenon is bound to occur as the policy drifts, unless the training data is trivial for the model or has already been learned.

File: verl/workers/rollout/vllm_rollout/utils.py [hunk 1]
```diff
@@ -176,7 +176,13 @@ def monkey_patch_model(self, vocab_size: int):
         # patch weight loader to support MoE model
         patch_vllm_moe_model_weight_loader(self.model_runner.model)
 
-    def update_weights_from_ipc(self, peft_config: dict = None, base_sync_done=False, use_shm: bool = False):
+    def update_weights_from_ipc(
+        self,
+        peft_config: dict = None,
+        base_sync_done=False,
+        use_shm: bool = False,
+        use_speculative_decoding: bool = False,
```
Comment (ArronHZG) [inline verl/workers/rollout/vllm_rollout/utils.py:184]: If we only transmit whether to use_speculative_decoding, it seems unnecessary.

File: verl/workers/rollout/vllm_rollout/utils.py [hunk 2]
```diff
@@ -216,10 +222,27 @@ def update_weights_from_ipc(self, peft_config: dict = None, base_sync_done=False
...
         receiver.receive_weights(
             on_bucket_received=lambda weights: self._update_weights(
-                weights, peft_config=peft_config, base_sync_done=base_sync_done
+                weights,
+                peft_config=peft_config,
+                base_sync_done=base_sync_done,
             )
         )
 
+        if use_speculative_decoding:
```
Comment (ArronHZG) [inline verl/workers/rollout/vllm_rollout/utils.py:231]: I understand that speculative_decoding will be executed multiple times here.

**Comment by ArronHZG** [comment]
> > DAPO-Math-17k
> 
> When using the DAPO-Math-17k dataset, we observed a reward collapse around step 100-120. If you try training for more steps, this phenomenon is bound to occur as the policy drifts, unless the training data is trivial for the model or has already been learned.

Is the MTP model or the Draft model being used?
Which specific base model is being used?

**Comment by tpx818** [comment]
> > > DAPO-Math-17k
> > 
> > 
> > When using the DAPO-Math-17k dataset, we observed a reward collapse around step 100-120. If you try training for more steps, this phenomenon is bound to occur as the policy drifts, unless the training data is trivial for the model or has already been learned.
> 
> Is the MTP model or the Draft model being used? Which specific base model is being used?

We are using the Eagle3 model, with Qwen3-8B.

**Comment by chrisociepa** [comment]
Have you tested it with FSDP2 instead of Megatron?

**Comment by aoshen02** [comment]
The draft weight update is merged in vllm 46725
